gammapy/utils/fitting/sampling.py

```diff
@@ -0,0 +1,181 @@
+    Parameter limits should be done here through uniform prior ditributions
+    """
+    logprob = 0
+    for par in dataset.parameters.free_parameters:
+        logprob += uniform_prior(par.value, par.min, par.max)
+
+    return logprob
+
+
+def lnprob(pars, dataset):
+    """Estimate the likelihood of a model including prior on parameters."""
+    # Update model parameters factors inplace
+    for factor, par in zip(pars, dataset.parameters.free_parameters):
+        par.factor = factor
+
+    lnprob_priors = ln_uniform_prior(dataset)
+
+    # dataset.likelihood returns Cash statistics values
+    # emcee will maximisise the LogLikelihood so we need -dataset.likelihood
+    total_lnprob = -dataset.likelihood() + lnprob_priors
+
+    return total_lnprob
+
+
+def run_mcmc(dataset, nwalkers=8, nrun=1000, threads=1):
+    """
+    Run the MCMC sampler.
+
+    Parameters
+    ----------
+    dataset : `gammapy.utils.fitting.Dataset`
+        A gammapy dataset object. This contains the observed counts cube,
+        the exposure cube, the psf cube, and the sky model and model.
+        Each free parameter in the sky model is considered as parameter for the MCMC.
+    nwalkers: int
+        Required integer number of walkers to use in ensemble.
+        Minimum is 2*nparam+2, but more than that is usually better.
+        Must be even to use MPI mode.
+    nrun: int
+        Number of steps for walkers. Typically at least a few hundreds (but depends on dimensionality).
+        Low nrun (<100?) will underestimate the errors.
+        Samples that would populate the distribution are nrun*nwalkers.
+        This step can be ~seen as the error estimation step.
+    threads: (optional)
+        The number of threads to use for parallelization. If ``threads == 1``,
+        then the ``multiprocessing`` module is not used but if
+        ``threads > 1``, then a ``Pool`` object is created and calls to
+        ``lnpostfn`` are run in parallel.
+
+    Returns
+    -------
+    sampler : `emcee.EnsembleSampler`
+        sampler object containing the trace of all walkers.
+    """
+    dataset.parameters.autoscale()  # Autoscale parameters
+    pars = dataset_free_parameter_factors(
+        dataset
+    )  # get a tuple of free parameters from dataset
+    ndim = len(pars)
+
+    # Initialize walkers in a ball of relative size 0.5% in all dimensions if the
+    # parameters have been fit, or to 10% otherwise
+    # TODO: the spread of 0.5% below is valid if a pre-fit of the model has been obtained.
+    # currently the run_mcmc() doesn't know the status of previous fit.
+    spread = 0.5 / 100
+    p0var = np.array([spread * pp for pp in pars])
+    p0 = emcee.utils.sample_ball(pars, p0var, nwalkers)
+
+    labels = []
+    for par in dataset.parameters.free_parameters:
+        labels.append(par.name)
+        if (par.min is np.nan) and (par.max is np.nan):
+            logging.warning(
```

Comment:
Put `log.info` instead of `logging.info`. This will use the `log` object that you created at the top of this module. That's the coding pattern we and many libraries like Astropy use for logging.

Reply:
fixed